File: packages/coast_audio_miniaudio/lib/src/ma_bridge/filters/mab_high_shelf_filter.dart
import 'dart:ffi';

import 'package:coast_audio/coast_audio.dart';
import 'package:coast_audio_miniaudio/generated/ma_bridge_bindings.dart';
import 'package:coast_audio_miniaudio/src/ma_bridge/filters/mab_filter_base.dart';
import 'package:coast_audio_miniaudio/src/ma_extension.dart';

/// High shelf filter. It uses ma_hishelf2 implementations.
class MabHighShelfFilter extends MabFilterBase {
  MabHighShelfFilter({
    required this.format,
    required double gainDb,
    required double shelfSlope,
    required double frequency,
    super.memory,
  })  : _gainDb = gainDb,
        _shelfSlope = shelfSlope,
        _frequency = frequency {
    final config = library.mab_high_shelf_filter_config_init(format.sampleFormat.mabFormat.value, format.sampleRate, format.channels, gainDb, shelfSlope, frequency);
    library.mab_high_shelf_filter_init(_pHPF, config).throwMaResultIfNeeded();
    _updateLatency();
  }

  final AudioFormat format;

  double _gainDb;
  double get gainDb => _gainDb;

  double _shelfSlope;
  double get shelfSlope => _shelfSlope;

  double _frequency;
  double get frequency => _frequency;

  late AudioTime _latency;
  AudioTime get latency => _latency;

  late final _pHPF = allocate<mab_high_shelf_filter>(sizeOf<mab_high_shelf_filter>());

  @override
  void process(AudioBuffer bufferOut, AudioBuffer bufferIn) {
    assert(bufferOut.sizeInFrames >= bufferIn.sizeInFrames);
    library.mab_high_shelf_filter_process(_pHPF, bufferOut.pBuffer.cast(), bufferIn.pBuffer.cast(), bufferIn.sizeInFrames).throwMaResultIfNeeded();
  }

  /// Reinit the filter parameters while keeping internal state.
  void update({double? gainDb, double? shelfSlope, double? frequency}) {
    final config = library.mab_high_shelf_filter_config_init(
      format.sampleFormat.mabFormat.value,
      format.sampleRate,
      format.channels,
      gainDb ?? this.gainDb,
      shelfSlope ?? this.shelfSlope,
      frequency ?? this.frequency,
    );
    library.mab_high_shelf_filter_reinit(_pHPF, config).throwMaResultIfNeeded();
    _gainDb = gainDb ?? this.gainDb;
    _shelfSlope = shelfSlope ?? this.shelfSlope;
    _frequency = frequency ?? this.frequency;
    _updateLatency();
  }

  void _updateLatency() {
    final frameCount = library.mab_high_shelf_filter_get_latency(_pHPF);
    _latency = AudioTime.fromFrames(frames: frameCount, format: format);
  }

  @override
  void uninit() {
    library.mab_high_shelf_filter_uninit(_pHPF);
  }
}

File: packages/coast_audio_miniaudio/mabridge/src/mab_high_shelf_filter.c
#define MINIAUDIO_IMPLEMENTATION
#include "miniaudio.h"

#include "mab_types.h"
#include "mab_high_shelf_filter.h"

mab_high_shelf_filter_config mab_high_shelf_filter_config_init(mab_format format, u_int32_t sampleRate, u_int32_t channels, double gainDb, double shelfSlope, double frequency)
{
  mab_high_shelf_filter_config config = {
      .format = format,
      .sampleRate = sampleRate,
      .channels = channels,
      .gainDb = gainDb,
      .shelfSlope = shelfSlope,
      .frequency = frequency,
  };
  return config;
}

mab_result mab_high_shelf_filter_init(mab_high_shelf_filter* pHSF, mab_high_shelf_filter_config config)
{
  ma_hishelf2_config maConfig = ma_hishelf2_config_init(mab_cast(ma_format, config.format), config.channels, config.sampleRate, config.gainDb, config.shelfSlope, config.frequency);
  ma_hishelf2* pData = MAB_MALLOC(sizeof(ma_hishelf2));
  ma_result result = ma_hishelf2_init(&maConfig, NULL, pData);
  if (result != MA_SUCCESS) {
    MAB_FREE(pData);
    return result;
  }

  pHSF->pData = pData;
  return MA_SUCCESS;
}

mab_result mab_high_shelf_filter_process(mab_high_shelf_filter* pHSF, void* pFramesOut, const void* pFramesIn, u_int64_t frameCount)
{
  ma_hishelf2 *pData = (ma_hishelf2*)pHSF->pData;
  return ma_hishelf2_process_pcm_frames(pData, pFramesOut, pFramesIn, frameCount);
}

mab_result mab_high_shelf_filter_reinit(mab_high_shelf_filter* pHSF, mab_high_shelf_filter_config config)
{
  ma_hishelf2 *pData = (ma_hishelf2*)pHSF->pData;
  ma_hishelf2_config maConfig = ma_hishelf2_config_init(mab_cast(ma_format, config.format), config.channels, config.sampleRate, config.gainDb, config.shelfSlope, config.frequency);
  return ma_hishelf2_reinit(&maConfig, pData);
}

u_int32_t mab_high_shelf_filter_get_latency(mab_high_shelf_filter* pHSF)
{
  ma_hishelf2 *pData = (ma_hishelf2*)pHSF->pData;
  return ma_hishelf2_get_latency(pData);
}

void mab_high_shelf_filter_uninit(mab_high_shelf_filter* pHSF)
{
  ma_hishelf2 *pData = (ma_hishelf2*)pHSF->pData;
  ma_hishelf2_uninit(pData, NULL);
  MAB_FREE(pData);
  pHSF->pData = NULL;
}
